Fixed viscoelastic stress output

tiannh7 opened the pull request:
## Description of Changes
This PR fixes an issue related to the viscoelastic stress output. Previously, there were inconsistencies in the sign convention and incorrect values in certain cases. The changes ensure that the stress components are correctly calculated and follow the expected conventions.

The attached images illustrate the differences before and after the fix. The previous implementation showed incorrect stress distribution, while the new implementation correctly reflects the expected physical behavior.

Since I do not frequently submit PRs, please feel free to point out any formatting or procedural errors. I appreciate any feedback!

![Before Fix](https://github.com/user-attachments/assets/bb192054-76f6-45ca-8d61-39ccdc606390)
![After Fix](https://github.com/user-attachments/assets/dc2d4073-1933-4fb5-9adb-e71c9a694ebb)

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented:
Hi @tiannh7 and thankyou for opening this pull request! Everything you did seems correct, this was a bug. 

I just started our testers and I would expect some results to change, I will let you know if that is the case and then I can show you how to update the results.

Just for future reference: Can you confirm that this is essentially the same change we had to do in #5480 and #4375? We had some issues with the sign convention of stress in the past, and this looks like it is the next place where the same problem appears.

gassmoeller commented:
/rebuild

tiannh7 commented:
Hi @gassmoeller,

Thank you for reviewing my PR and for starting the testers! I really appreciate your help.

Yes, my changes are indeed consistent with what was done in #5480 and #4375. This issue also stems from the sign convention problem, and I followed the correct convention as described in other references to fix it in this file.

I hadn’t noticed those two PRs before, but looking at them now, I see that the fixes are quite similar. Thank you for pointing them out! Please let me know if there are any additional adjustments needed.

Thanks again for your guidance!

gassmoeller commented:
Ok, thanks for checking :+1:. I have two small requests before we can merge this PR:
1. Since this fixes an actual bug, can you add a changelog entry like the ones added in #5480 and #4375? This essentially just means adding a file into the folder `doc/modules/changes` like the ones that are already there. Modify name and content of the file to fit the current date and your name and describe what you fixed (you can make it very similar to the files in the other PRs). This will ensure users will learn about this bugfix and also make sure you get credit for the fix.
2. It looks like all the tests pass, which in this case is a bad sign, because it means the behavior of these postprocessors is not tested in models with elasticity at the moment. Can you either add a new test that uses the postprocessors and elasticity, or take an existing test and modify that one? The process of adding tests is described here: https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html. I think a good candidate would be the test [tests/viscoelastic_stress_build-up.prm](https://github.com/geodynamics/aspect/pull/4375/files#diff-9592f48feaafa2d214c818ac2a6e31ed95afef838cc33f4a1d7ca7c40d73a99c). You could copy that test and its output directory, give it a new name, and add the postprocessors you have changed in this PR. Then update this PR to let our testers run again, and use the tester output to update the results as described in the documentation I linked above. If you get stuck at any point let me know here and I can walk you through the process.

tiannh7 commented:
Hi @gassmoeller,

I have added the changelog entry and updated the test outputs as requested. Let me know if there’s anything else I should adjust. 

Thanks!

gassmoeller commented:
Hi @tiannh7, almost there. There are 2 remaining small issues:
1. The [linux/indent](https://github.com/geodynamics/aspect/actions/runs/13990929034/job/39177548983?pr=6261) tester is complaining that there are several spaces at the end of lines in the changelog entry. You can read [here](https://aspect-documentation.readthedocs.io/en/latest/user/extending/contributing.html#coding-conventions) how to fix that (and any other indentation issues). Please follow the description, run `make indent` and then update the PR.
2. For your new test: Please update the description of the test at the beginning of the parameter file. For your purpose this could be as simple as saying: `This is a copy of the test viscoelastic_stress_build-up, which additionally tests the output of the viscoelastic stress postprocessors.` Additionally the output of our reference tester is marginally different from the output you included. This is completely normal and could be caused because the tester runs a different operating system or different deal.II version than you used. A description for how to fix this is [at the bottom of this page](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html).

Let me know when you have fixed these two issues or in case you have questions, afterwards I can merge the fix.

tiannh7 commented:
Hi @gassmoeller,  

Thank you for your detailed guidance! I have updated the changelog and the description in the parameter file as suggested.  

However, I encountered a issue that I would like to clarify:  

**Formatting with `make indent`**  
   The required `astyle` version is 2.04, but on macOS, I can only install version 3.6.7. I ran `make indent` with the available version, but I am not sure if the formatting fully matches the expected output. Would it be acceptable, or should I run it on a Linux system with `astyle 2.04`?  

Thank you again for your help! I appreciate your time and guidance.

gassmoeller commented:
Hi @tiannh7, thanks everything is in order now. I will merge this PR, congratulations to your first contribution to ASPECT! :tada:

For future PRs: The version of astyle is important, because there are small differences in code style between the versions. It doesnt matter for this PR, because your changes were relatively simple and so the tester passed. But you do not need a linux system for that, astyle version 2.04 should work on Macos just fine, you may just need to install it yourself. E.g. `mkdir astyle && cd astyle && wget 'https://sourceforge.net/projects/astyle/files/astyle/astyle 2.04/astyle_2.04_linux.tar.gz' && tar -zxvf astyle_2.04_linux.tar.gz && cd astyle/build/gcc && make && sudo make install` should install it for everyone on your system. Careful this may overwrite your current astyle, instead you can specify an install directory before running make install, just make sure your system finds the executable when you run `astyle` in a terminal. Or you could try using a package manager that can install version 2.04 (I dont know if homebrew or one of the other managers support versions that old). Another option could be to use our reference tester docker container [geodynamics/aspect-tester:focal-dealii-9.5-v3](https://hub.docker.com/layers/geodynamics/aspect-tester/focal-dealii-9.5-v3/images/sha256-69d1ae23e9abc6d76cc33e4b701dee2930bccdfeafd0c2c5c8bd6ff0feca7a24). It is publicly available and has the correct version of astyle installed.
